File: script/transfer.lua
local outKey = KEYS[1]
local inKey = KEYS[2]
local outField = ARGV[1]
local inField = ARGV[2]
local amount = tonumber(ARGV[3])
local initValue = ARGV[4]
-- 检查转出账户是否存在
if redis.call("HSETNX", outKey, outField, initValue) == 1 then
    return { 201, "out" }
end
-- 获取转出账户
local outResult = redis.call("HGET", outKey, outField)
-- 初始化中等待
if outResult == initValue then
    return { 202, "out" }
end
-- 初始化
if outResult == false then
    return { 201, "out" }
end
-- 检查转入账户是否存在
if redis.call("HSETNX", inKey, inField, initValue) == 1 then
    return { 201, "in" }
end
-- 获取转入账户
local inResult = redis.call("HGET", inKey, inField)
-- 初始化中等待
if inResult == initValue then
    return { 202, "in" }
end
if inResult == false then
    return { 201, "in" }
end
-- 转换为数字
local outBefore = tonumber(outResult)
local inBefore = tonumber(inResult)
-- 返回当前余额
if amount == 0 then
    return { 200, "Success", outBefore, inBefore, outBefore, inBefore }
end
-- 余额不足
if outBefore - amount < 0 then
    return { 203, "Balance insufficient", outBefore, inBefore, outBefore, inBefore }
end
-- 原子转账：扣
local outBalance = redis.call("HINCRBY", outKey, outField, -amount)
-- 原子转账： 加
local inBalance = redis.call("HINCRBY", inKey, inField, amount)
return { 200, "Success", outBefore, inBefore, tonumber(outBalance), tonumber(inBalance) }

File: src/tips/Transfer.php
<?php

namespace AlonePhp\Redis\tips;
/**
 * - 双帐户转帐
 * - 状态码说明
 * - 200: 成功
 * - 201: 需要初始化
 * - 202: 初始化中等待
 * - 203: 余额不足
 * - 204: 执行超时
 * - 205: 异常错误
 */
class Transfer {
    public int        $code       = 0;   // 状态码
    public string     $msg        = "";  // 提示信息
    public float|int  $amount     = 0;   // 操作额度
    public string     $outKey     = "";  // 转出key
    public string     $outField   = "";  // 转出字段
    public float|int  $outBefore  = 0;   // 转出前额度
    public float|int  $outBalance = 0;   // 转出后额度
    public string     $inKey      = "";  // 转入key
    public string     $inField    = "";  // 转入字段
    public float|int  $inBefore   = 0;   // 转入前额度
    public float|int  $inBalance  = 0;   // 转入后额度
    public float|int  $execute    = 0;   // 执行时间
    public array|null $error      = [];  // 报错详细

    public function __construct(array $res) {
        foreach (array_intersect_key($res, get_object_vars($this)) as $k => $v) {
            $this->$k = $v;
        }
    }

    /**
     * 判断是否转帐成功
     * @return bool
     */
    public function is(): bool {
        return $this->code == 200 && $this->isOut() && $this->isIn();
    }

    /**
     * 判断转出是否成功
     * @return bool
     */
    public function isOut(): bool {
        return $this->outBefore - $this->outBalance == $this->amount;
    }

    /**
     * 判断转入是否成功
     * @return bool
     */
    public function isIn(): bool {
        return $this->inBalance - $this->inBefore == $this->amount;
    }
}